snakemake/remote/GS.py

```diff
@@ -95,12 +126,30 @@ def size(self):
             return self._iofile.size_local
 
     def download(self):
-        if self.exists():
-            os.makedirs(os.path.dirname(self.local_file()), exist_ok=True)
-            self.blob.download_to_filename(self.local_file())
+        if not self.exists():
+            return None
+
+        # Create the directory for the intended file
+        os.makedirs(os.path.dirname(self.local_file()), exist_ok=True)
+        checksums_match = False
+
+        # Continue trying to download until hash matches
+        while not checksums_match:
+
+            # ideally we could calculate hash while streaming to file with provided function
+            # https://github.com/googleapis/python-storage/issues/29
+            with open(self.local_file(), "wb") as blob_file:
+                parser = Crc32cCalculator(blob_file)
+                self.blob.download_to_file(parser)
             os.sync()
-            return self.local_file()
-        return None
+
+            # Compute local hash and verify correct
+            if parser.hexdigest() != self.blob.crc32c:
+                os.remove(self.local_file())
```

Comment:
A final thought: in theory, this could go on forever. I think we should give it 3 tries, and then raise a `WorkflowError("Failed to download {} from remote: checksums did not match 3 times.".format(self.remote_file()))`.

Reply:
okay, added! I made `retry_count` a variable so (at some future point) it might be exposed. If some kind of issue comes up with requests being too quickly one after the other, we could add a delay (e.g., exponential or some kind of backoff) but I don't see a strong need for this until there is evidence for it.